OSDOCS-1855: OVN Egress Router

mikemckiernan opened the pull request:
After helpful comments from Daniel, there are substantial changes on two HTML pages:

https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/deploying-egress-router-ovn-redirection.html

...and the page that precedes it:

https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/using-an-egress-router-ovn.html

...and ER is set as "Supported" instead of "Partial support" in the [support matrix](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/about-ovn-kubernetes.html#nw-ovn-kubernetes-matrix_about-ovn-kubernetes).

The support matrix is also visible from the [OpenShift SDN section](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/openshift_sdn/about-openshift-sdn.html#nw-ovn-kubernetes-matrix_about-openshift-sdn) that switches the column sequence.

On the OpenShift SDN to OVN-K migration page, there is a section for [Egress router](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.html#egress-router-pods_migrate-from-openshift-sdn).

----
applies to enterprise-4.8 and milestone Future Release.

mikemckiernan commented on modules/nw-egress-router-cr.adoc:
```diff
@@ -0,0 +1,118 @@
+    fallbackIP: "<egress_destination>"  <.>
+  }
+----
+<.> The `addresses` field specifies the IP addresses to configure on the secondary network interface.
```
I found the field descriptions from the [crd](https://github.com/openshift/cluster-network-operator/blob/master/manifests/0000_70_cluster-network-operator_01_egr_crd.yaml) in the CNO repo.  Please check.

mikemckiernan commented on modules/nw-egress-router-redirect-mode-ovn.adoc:
```diff
@@ -3,11 +3,11 @@
 // * networking/ovn_kubernetes_network_provider/deploying-egress-router-ovn-redirection.adoc
 
```
@jboxman , I removed "pod" from the text because this work introduces a dedicated custom resource.  If removing pod was a mistake, just let me know and thanks.

danielmellado replied:
++, also, this is now deployed using a 'deployment' rather than a specific pod resource, just in case you'd like to mention it.

mikemckiernan commented on modules/nw-egress-router-redirect-mode-ovn.adoc:
```diff
@@ -16,9 +16,7 @@ After you add an egress router pod, the client pods that need to use the reserve
-. Create an egress router pod.
+. Create an egress router definition.
 
 . To ensure that other pods can find the IP address of the egress router pod, create a service that uses the egress router pod, as in the following example:
```
@danielmellado, @weliang1, is this suggestion to add a service still plausible?

mikemckiernan commented on modules/nw-egress-router-redirect-mode-ovn.adoc:
```diff
@@ -30,46 +28,77 @@ metadata:
   type: ClusterIP
   selector:
-    name: egress-router-pod
+    app: egress-router-cni
```
@danielmellado, please confirm or correct me on the correct way to associate the service with the router.  (I need to state the label that is automatically assigned clearly in this section or another.)

danielmellado replied:
Yeah, you can check an example of this in https://etherpad.opendev.org/p/egress-router-cni-debug

danielmellado commented on modules/nw-egress-router-cr.adoc:
```diff
@@ -0,0 +1,118 @@
+
+<.> The `protocol` field supports TCP, UDP, or SCTP.
+
+<.> Optional: The `fallbackIP` field specifies one IP address for a remote destination. If redirect rules are not specified, all traffic from the router is redirected to this destination. If redirect rules are specified, any connections to the egress router on any port that is not defined in a rule is redirected to this destination. If redirect rules are specified and no fallback IP address is pr ...
```
I'd remove fallbackIP for now, this was added to the CRD, but won't be funcional for now.

mikemckiernan commented:
/assign @danielmellado 
/assign @weliang1

For the newest material, I added some comments and questions files that I modified. The HTML is available on the "[Deploying an egress router pod in redirect mode](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/deploying-egress-router-ovn-redirection.html)" page.

Some doubts:
* [ ] This [Egress router pod implementation](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/using-an-egress-router-ovn.html#nw-egress-router-about-router-pod-implementation_using-an-egress-router-ovn) heading is a candidate location for stating the static information--the label that the controller adds to the pod, the name of the NAD, the name of the pod, the name of the deployment, ...more?
* [ ] A little further down on that page is the [Failover configuration](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/using-an-egress-router-ovn.html#nw-egress-router-about-failover_using-an-egress-router-ovn) heading.  A deployment is automatically added.  I can revise this section to indicate that a deployment is added (with a specific name).  Regarding the `oc expose...` command suggestion, would the syntax be `oc expose deployment/egress-router-cni-deployment --port 8443,8080`?  I have no idea how to perform the simplest of checks on that.
* [ ] What else have I forgotten?

As always, thanks folks.

danielmellado commented:
> /assign @danielmellado
> /assign @weliang1
> 
> For the newest material, I added some comments and questions files that I modified. The HTML is available on the "[Deploying an egress router pod in redirect mode](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/deploying-egress-router-ovn-redirection.html)" page.
> 
> Some doubts:
> 
>     * [ ]  This [Egress router pod implementation](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/using-an-egress-router-ovn.html#nw-egress-router-about-router-pod-implementation_using-an-egress-router-ovn) heading is a candidate location for stating the static information--the label that the controller adds to the pod, the name of the NAD, the name of the pod, the name of the deployment, ...more?

I think that's mostly it! Thanks for catching this up!
> 
>     * [ ]  A little further down on that page is the [Failover configuration](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/using-an-egress-router-ovn.html#nw-egress-router-about-failover_using-an-egress-router-ovn) heading.  A deployment is automatically added.  I can revise this section to indicate that a deployment is added (with a specific name).  Regarding the `oc expose...` command suggestion, would the syntax be `oc expose deployment/egress-router-cni-deployment --port 8443,8080`?  I have no idea how to perform the simplest of checks on that.
So regarding failover configuration, you can mention what is written on the old documentation. This is now been done in an automatic way by the controller. Regarding the command, IMHO you're mostly right, I'll let Weibin confirm, but you would have a
`oc expose deployment/egress-router-cni-deployment --port 8443` and then you would have to create another one, with a different name, such as 
`oc expose deployment/egress-router-cni-deployment --port 8080 --name foo` or otherwise it will conflict. Hint [1]

[1] https://stackoverflow.com/questions/54402920/how-to-expose-2-ports-in-openshift-for-jmeter-client
> 
>     * [ ]  What else have I forgotten?
> 

I would mention that when creating the object, it will create the associated NAD and Deployment for you, and that if you delete the EgressRouter, they will also get wiped by the controller.
> 
> As always, thanks folks.

Thanks!

Daniel

danielmellado commented:
should be ok as you would need a baremetal for the macvlan to work, but
I'll double check

El lun., 26 abr. 2021 21:50, mikemckiernan ***@***.***>
escribió:

> @danielmellado <https://github.com/danielmellado> , @weliang1
> <https://github.com/weliang1>, please peek at the updates to the
> "Failover configuration" heading. I added sample YAML for a service, but
> *I* wasn't able to route traffic from a test pod (with curl) to the
> service IP and get anything back. As long as it's just my cooties and not
> that the suggestion is wrong, that's OK.
>
>
> https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/using-an-egress-router-ovn.html#nw-egress-router-about-failover_using-an-egress-router-ovn
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/openshift-docs/pull/31896#issuecomment-827100114>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAVQ5VWKQEMHAOO74BKEDTKW7YJANCNFSM43PJHNPA>
> .
>

mikemckiernan commented:
Daniel, that was my mistake, I wasn't on bare metal.

/hold cancel

@weliang1 PTAL

weliang1 commented:
> Daniel, that was my mistake, I wasn't on bare metal.
> 
> /hold cancel
> 
> @weliang1 PTAL

@mikemckiernan I need run a quick test after https://bugzilla.redhat.com/show_bug.cgi?id=1951158 got fixed, then I can comment this PR based on my quick testing results.

jboxman commented:
@mikemckiernan, I'd swap [2] and [3] so in the table it flows as 1, 2, 3. Right now, it reads as 1, 3, 2 from left to right, top to bottom.

mikemckiernan commented:
/assign @msherif1234

Mohamed, I added `fallbackIP` based on an update to CNO.  It's on the [Deploying an egress router in redirect mode](https://deploy-preview-31896--osdocs.netlify.app/openshift-enterprise/latest/networking/ovn_kubernetes_network_provider/deploying-egress-router-ovn-redirection.html#nw-egress-router-ovn-cr_deploying-egress-router-ovn-redirection) page.

I got disoriented with what needed to be an IP and what needed to be a CIDR.  I fiddled with a cluster that I tried to start with the CNO#1094 PR and my conclusion--based on trial and error--is that both `spec.redirect.fallbackIP` and `spec.addresses.ip` need to be IP addresses and enclosed in quotation marks (aka, a string).

```
The EgressRouter "egress-router-redirect" is invalid:                                                                                                    [54/18810]
* <nil>: Invalid value: "": "spec.redirect.fallbackIP" must validate at least one schema (anyOf)
* spec.redirect.fallbackIP: Invalid value: "192.168.1.5/32": spec.redirect.fallbackIP in body must be of type ipv4: "192.168.1.5/32"
* <nil>: Invalid value: "": "spec.addresses.ip" must validate at least one schema (anyOf)
* spec.addresses.ip: Invalid value: "192.168.3.99/32": spec.addresses.ip in body must be of type ipv4: "192.168.3.99/32"
```

jboxman commented:
@mikemckiernan, I'm still erroneously claiming that this isn't supported in this doc: https://docs.openshift.com/container-platform/4.8/networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.html#considerations-migrating-ovn-kubernetes-network-provider_migrate-from-openshift-sdn

It's probably worth updating this:

> ## Egress router pods
> OVN-Kubernetes does not support using egress router pods in OpenShift Container Platform 4.8.

jboxman commented on modules/nw-egress-router-about.adoc:
```diff
@@ -134,36 +142,38 @@ spec:
 endif::openshift-sdn[]
 
 ifdef::ovn[]
+To avoid downtime, the Cluster Network Operator deploys the egress router pod as a deployment resource. The deployment name is `egress-router-cni-deployment`.  The pod that corresponds to the deployment has a label of `app=egress-router-cni`.
```
```suggestion
To avoid downtime, the Cluster Network Operator deploys the egress router pod as a deployment resource. The deployment name is `egress-router-cni-deployment`. The pod that corresponds to the deployment has a label of `app=egress-router-cni`.
```

jboxman commented on modules/nw-egress-router-redirect-mode-ovn.adoc:
```diff
@@ -16,11 +16,9 @@ After you add an egress router pod, the client pods that need to use the reserve
 .Procedure
 
-. Create a network attachment definition.
+. Create an egress router definition.
```
Does it matter in what namespace this happens? Is EgressRouter cluster scoped?

mikemckiernan replied:
The namespace matters.  I've updated the YAML for the egress router def to show that it is possible to use `metadata.namespace`.  I used "Configuration for bridge" in the Multus area for the lingo.

jboxman reviewed (commented):
Looking really excellent.

weliang1 commented on modules/nw-egress-router-network-attachment.adoc:
```diff
@@ -1,50 +0,0 @@
-// Module included in the following assemblies:
```
Supporting egressrouter CR in v4.8, my understanding is we do not need manually setup  NetworkAttachmentDefinition object for egress router pod any more

mikemckiernan replied:
Agreed.  This file is deleted for 4.8.

ahardin-rh commented on modules/nw-ovn-kubernetes-migration-about.adoc:
```diff
@@ -95,7 +95,9 @@ For more information on using an egress firewall in OVN-Kubernetes, see "Configu
-OVN-Kubernetes does not support using egress router pods in {product-title} {product-version}.
+OVN-Kubernetes supports egress router pods in redirect mode. OVN-Kubernetes does not support egress router pods in HTTP proxy mode or DNS proxy mode.
+
+When you deploy an egress router with the Cluster Network Operator, as shown in the documentation, you cannot specify a node selector to control which node is used to host the egress router pod.
```
I don't think "as shown in the documentation" is  necessary here

ahardin-rh commented:
Just one minor comment from me. Looks good!

mikemckiernan commented:
/cherrypick enterprise-4.8
